File: src/core/log-volume.ts
import { classifyLogSeverity } from "./log-store.js";
import type { MetricSample } from "./metrics-store.js";
import type { LogEntry } from "./types.js";

/**
 * How much a service said, and how badly, over the range its graph already
 * covers.
 *
 * The CPU and memory panes gave the log tail an axis; this gives that axis the
 * one series it was still missing — the output itself. A CPU spike says *when*
 * something happened, but not whether the service noticed; a burst of stderr in
 * the same minute says it did, and points at the lines directly below.
 *
 * **Bucketed here rather than in the browser, because the browser cannot.** The
 * dashboard payload carries forty lines per service (`PAYLOAD_TAIL`), which is a
 * few seconds of a service mid-burst — far too short an axis to bucket. The ring
 * buffer this reads holds 500, and lives in the daemon. Sending 500 lines per
 * service per poll to let the client count them would be the same payload
 * mistake at ten times the size; sending sixty triples of small integers is not.
 */

export interface LogVolumeBucket {
  /** Start of the bucket, ms epoch — the same clock `MetricSample.t` is on. */
  t: number;
  /** Lines with nothing to declare. Carries the *volume* half of the signal. */
  info: number;
  warning: number;
  error: number;
}

/**
 * Sixty across the window, whatever the window is.
 *
 * A count, not a duration, because the range is whatever the metrics buffer
 * happens to hold: half an hour for a service that has been up all morning, two
 * minutes for one that just booted. Bucketing by a fixed *duration* would give
 * the young service four bars and the old one a thousand; bucketing by a fixed
 * count keeps the strip the same shape and lets the bar width mean "a sixtieth
 * of what you are looking at" in both.
 *
 * Sixty is also about the widest a panel can draw and still leave a bar with a
 * gap on each side of it at the narrow end of the layout.
 */
const BUCKET_COUNT = 60;

/**
 * Count each severity into evenly spaced buckets over the metric series' range.
 *
 * The range comes from `samples` rather than from the log lines themselves, and
 * that is the whole point: the strip is only honest if it is the *same* axis as
 * the panes above it. Deriving it from the lines would give the strip its own
 * ends, and two charts stacked with different x-ranges is a worse lie than no
 * chart at all — a spike would sit above a moment it did not happen at.
 *
 * Lines outside the range are dropped rather than clamped. A service that
 * emitted 400 lines at boot and has been quiet since is *quiet* in the window
 * being shown, and piling that history onto the first bucket would invent a
 * spike at a moment nothing happened.
 */
export function bucketLogVolume(
  entries: LogEntry[],
  samples: MetricSample[],
  bucketCount = BUCKET_COUNT,
): LogVolumeBucket[] {
  if (samples.length < 2) return [];
  const first = samples[0].t;
  const last = samples[samples.length - 1].t;
  const span = last - first;
  if (span <= 0) return [];

  const width = span / bucketCount;
  const buckets: LogVolumeBucket[] = Array.from({ length: bucketCount }, (_, index) => ({
    t: first + index * width,
    info: 0,
    warning: 0,
    error: 0,
  }));

  for (const entry of entries) {
    const at = Date.parse(entry.timestamp);
    if (Number.isNaN(at) || at < first || at > last) continue;
    /* The final bucket is closed at both ends so the newest line — which is
       exactly `last` whenever a sample and a line share a millisecond — lands in
       the strip rather than one index past it. */
    const index = Math.min(bucketCount - 1, Math.floor((at - first) / width));
    buckets[index][severityOf(entry)] += 1;
  }

  return buckets;
}

/**
 * Which band a line counts toward — decided by what it *says*, never by which
 * pipe it came down.
 *
 * The first version of this read `stream === "stderr"` first, on the reasoning
 * that the panel already counts a `STDERR` stat and paints those lines red in
 * the tail. Ten seconds of a real service settled it: of jobjourney-frontend's
 * ten stderr lines, **none** were errors. They were `[TypeScript] Found 0
 * errors.`, a Browserslist freshness notice, and six Tailwind lines beginning
 * with the literal word `warn`. A rule that paints "found 0 errors" red is not
 * strict, it is broken, and it costs the amber band the exact case amber is for.
 *
 * A red that fires on every dev server is a red nobody looks at, so the pipe is
 * not consulted at all. Nothing is lost: a service that dies on stderr dies
 * saying `panic`, `fatal`, `Traceback`, `uncaught`, `EADDRINUSE` or `error`, and
 * every one of those is what `classifyLogSeverity` is looking for.
 *
 * This does not contradict the `STDERR` counter beside it — that counter names
 * what it counts, and so do these bands. "10 stderr lines, 0 errors" is two true
 * statements about the same ten lines.
 */
function severityOf(entry: LogEntry): "info" | "warning" | "error" {
  return classifyLogSeverity(entry.text) ?? "info";
}

File: crates/nomoreide-core/src/log_volume.rs
//! How much a service said, and how badly, over the range its graph covers.
//!
//! The Rust half of `src/core/log-volume.ts`. The CPU and memory panes give the
//! log tail an axis; this gives that axis the series it was missing. A CPU spike
//! says *when* something happened but not whether the service noticed; a burst
//! of errors in the same minute says it did.
//!
//! **Bucketed here rather than in the browser, because the browser cannot.** The
//! dashboard payload carries forty lines per service, which is a few seconds of
//! a service mid-burst — far too short an axis to bucket. The ring buffer this
//! reads holds five hundred and lives in the daemon; sending all of them per
//! poll so a client could count them would be the same payload mistake at ten
//! times the size.

use serde_json::{json, Value};

/// Sixty across the window, whatever the window is.
///
/// A count, not a duration, because the range is whatever the metric buffer
/// happens to hold: half an hour for a service up all morning, two minutes for
/// one that just booted. A fixed duration would give the young service four
/// bars and the old one a thousand; a fixed count keeps the strip the same
/// shape and lets a bar mean "a sixtieth of what you are looking at" in both.
const BUCKET_COUNT: usize = 60;

/// Count each severity into evenly spaced buckets over the samples' range.
///
/// The range comes from the samples rather than from the lines, and that is the
/// point: the strip is only honest if it is the *same* axis as the panes above
/// it. Lines outside it are dropped rather than clamped — a service that
/// emitted four hundred lines at boot and has been quiet since is quiet in the
/// window being shown, and piling that history onto the first bucket would
/// invent a spike at a moment nothing happened.
pub fn bucket_log_volume(lines: &[(String, String)], sample_times: &[f64]) -> Vec<Value> {
    if sample_times.len() < 2 {
        return Vec::new();
    }
    let first = sample_times[0];
    let last = sample_times[sample_times.len() - 1];
    let span = last - first;
    if span <= 0.0 {
        return Vec::new();
    }

    let width = span / BUCKET_COUNT as f64;
    let mut counts = vec![[0_u64; 3]; BUCKET_COUNT];
    for (timestamp, text) in lines {
        let at = match chrono::DateTime::parse_from_rfc3339(timestamp) {
            Ok(parsed) => parsed.timestamp_millis() as f64,
            Err(_) => continue,
        };
        if at < first || at > last {
            continue;
        }
        // The final bucket is closed at both ends, so the newest line — which
        // is exactly `last` whenever a sample and a line share a millisecond —
        // lands in the strip rather than one index past it.
        let index = (((at - first) / width).floor() as usize).min(BUCKET_COUNT - 1);
        counts[index][band_of(text)] += 1;
    }

    counts
        .into_iter()
        .enumerate()
        .map(|(index, [info, warning, error])| {
            json!({
                "t": crate::js_number::value(first + index as f64 * width),
                "info": info,
                "warning": warning,
                "error": error,
            })
        })
        .collect()
}

/// Which band a line counts toward — decided by what it *says*, never by which
/// pipe it came down.
///
/// Reading the stream first looks right and is not: of one real dev server's
/// ten stderr lines, none were errors — they were a "found 0 errors" summary, a
/// freshness notice, and six lines beginning with the literal word `warn`. A
/// rule that paints "found 0 errors" red is not strict, it is broken, and it
/// costs the amber band the exact case amber is for. Nothing is lost: a service
/// that dies on stderr dies saying `panic`, `fatal` or `error`, and the
/// classifier is looking for all of them.
fn band_of(text: &str) -> usize {
    match crate::log_store::classify_severity(text).as_deref() {
        Some("error") => 2,
        Some("warning") => 1,
        _ => 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn line(timestamp: &str, text: &str) -> (String, String) {
        (timestamp.to_string(), text.to_string())
    }

    #[test]
    fn fewer_than_two_samples_have_no_range_to_bucket_over() {
        assert!(bucket_log_volume(&[], &[]).is_empty());
        assert!(bucket_log_volume(&[], &[1.0]).is_empty());
        // Two samples at the same instant span nothing.
        assert!(bucket_log_volume(&[], &[5.0, 5.0]).is_empty());
    }

    #[test]
    fn a_line_outside_the_window_is_dropped_rather_than_clamped() {
        let samples = [1_000.0, 61_000.0];
        let lines = [
            line("1970-01-01T00:00:00.500Z", "before the window"),
            line("1970-01-01T00:01:30.000Z", "after the window"),
            line("not a timestamp at all", "unreadable"),
        ];
        let buckets = bucket_log_volume(&lines, &samples);
        assert_eq!(buckets.len(), 60);
        let total: u64 = buckets
            .iter()
            .map(|bucket| {
                ["info", "warning", "error"]
                    .iter()
                    .map(|key| bucket[*key].as_u64().unwrap_or(0))
                    .sum::<u64>()
            })
            .sum();
        assert_eq!(total, 0);
    }

    #[test]
    fn the_last_bucket_is_closed_at_both_ends() {
        let samples = [0.0, 60_000.0];
        let lines = [line("1970-01-01T00:01:00.000Z", "exactly at the end")];
        let buckets = bucket_log_volume(&lines, &samples);
        assert_eq!(buckets[59]["info"].as_u64(), Some(1));
    }
}
